File: KWAudioUnit/KWAudioUnit/FRAudioUnit/FRMixerUnit.m
//
//  FRMixerUnit.m
//  TestCase_AudioController
//
//  Created by 상우 강 on 2016. 5. 27..
//  Copyright © 2016년 상우 강. All rights reserved.
//

#import "FRMixerUnit.h"



@interface FRMixerUnit () {
    
}

@end

@implementation FRMixerUnit

-(id)init {
    if (self = [super initWithType:kFRAudioUnitType_Mixer]) {
        OSStatus err = noErr;
        
        UInt32 busCount = MIXER_MAX_CHANNEL;
        UInt32 size = sizeof(busCount);
        AudioUnit audioUnit = [self audioUnit];
        err = AudioUnitSetProperty(audioUnit, kAudioUnitProperty_ElementCount, kAudioUnitScope_Input, 0, &busCount, size);
        if (err)
            NSLog(@"Mixer set MaximumChannel failed : %d", (int)err);
        
        [self setMixerVolume:0 withChannel:0];
    }
    return self;
}


-(void)setMixerVolume:(Float32)volume withChannel:(UInt32)channelNum {
    OSStatus err = noErr;
    AudioUnit audioUnit = [self audioUnit];
    err = AudioUnitSetParameter(audioUnit, kMultiChannelMixerParam_Volume, kAudioUnitScope_Input, channelNum, volume, 0);
    if (err)
        NSLog(@"mixer volume set failed : %d", (int)err);
}

-(void)setMixerOutputVolume:(Float32)volume {
    OSStatus err = noErr;
    err = AudioUnitSetParameter(self.audioUnit, kMultiChannelMixerParam_Volume, kAudioUnitScope_Output, 0, volume, 0);
    if (err)
        NSLog(@"mixer output volume set failed : %d", (int)err);
}

-(float)getMixerVolumeWithChannelNumber:(UInt32)channel {
    float returnValue = 0;
    OSStatus err = noErr;
    err = AudioUnitGetParameter(self.audioUnit, kMultiChannelMixerParam_Volume, kAudioUnitScope_Input, channel, &returnValue);
    if (err)
        NSLog(@"get mixer volume failed : %d", (int)err);
    return returnValue;
}

@end

File: Testing_iOS/Testing_iOS/lib/FRMixerUnit.swift
//
//  FRMixerUnit.swift
//  Testing_iOS
//
//  Created by exs-mobile 강상우 on 05/08/2019.
//  Copyright © 2019 강상우. All rights reserved.
//

import UIKit
import AudioToolbox

class FRMixerUnit: FRAudioUnit {
    let MIXER_MAX_CHANNEL   = 30
    let MIXER_AUDIO_CHANNEL = 1
    
    init() {
        super.init(.Mixer)
        
        var err = noErr
        var busCount = UInt32(MIXER_MAX_CHANNEL)
        let size = MemoryLayout.size(ofValue: busCount)
        
        err = AudioUnitSetProperty(self.audioUnit!, kAudioUnitProperty_ElementCount, kAudioUnitScope_Input, 0, &busCount, UInt32(size))
        if err != noErr {
            printError("Mixer Unit set Input Channel Count", err)
            return
        }
        
        for i in 0..<busCount {
            setMixerVolume(0, channel: i)
        }
    }
    
    
    //MARK: - Public
    public func setMixerVolume(_ volume: Float32, channel: UInt32) {
        var err = noErr
        err = AudioUnitSetParameter(self.audioUnit!, kMultiChannelMixerParam_Volume, kAudioUnitScope_Input, channel, volume, 0)
        if err != noErr {
            printError("Set Mixer Volume", err)
            return
        }
    }
    
    public func setMixerOutputVolume(_ volume: Float32) {
        var err = noErr
        err = AudioUnitSetParameter(self.audioUnit!, kMultiChannelMixerParam_Volume, kAudioUnitScope_Output, 0, volume, 0)
        if err != noErr {
            printError("set Mixer Output Volume", err)
            return
        }
    }
}
